modules/tensor_mechanics/src/materials/SingleVariableReturnMappingSolution.C

```diff
@@ -102,38 +130,65 @@ SingleVariableReturnMappingSolution::returnMappingSolve(const Real effective_tri
                                                         Real & scalar,
                                                         const ConsoleStream & console)
 {
-  std::stringstream iter_output;
-  std::stringstream * iter_output_ptr = (_output_iteration_info ? &iter_output : nullptr);
+  // construct the stringstream here only if the debug level is set to ALL
+  std::stringstream * iter_output =
+      (_debug_level == DebugLevel::ALL) ? new std::stringstream : nullptr;
 
   if (!_legacy_return_mapping)
   {
-    if (!internalSolve(effective_trial_stress, scalar, iter_output_ptr))
+    // do the internal solve and capture iteration info during the first round
+    // iff full history output is requested regardless of whether the solve failed or succeeded
+    if (!internalSolve(
+            effective_trial_stress, scalar, _output_iteration_info ? iter_output : nullptr))
     {
-      if (iter_output_ptr)
-        throw MooseException(iter_output_ptr->str());
+      // if full history output is only requested for failed solves we have to repeat
+      // the solve a second time
+      if (_debug_level == DebugLevel::ERROR && _output_iteration_info)
+      {
+        // construct string stream and capture iteration history in the second solve
+        iter_output = new std::stringstream;
+        internalSolve(effective_trial_stress, scalar, iter_output);
+        outputIterationSummary(iter_output, _iteration);
+        throw MooseException(iter_output->str());
+      }
+      else if (_debug_level == DebugLevel::NONE)
+        throw MooseException("");
```

Comment:
no exception message at all for a failed solve? Might it be better to have all exception calls at least say something like "Internal solve failed for material ", _name

Reply:
Well, if the user demands **silence** she or he gets silence :-)

Reply:
Not necessarily true. This user would want to know why the exception occurred, but not all the gory iteration details that goes along.

Reply:
Then the user should set `output_iteration_info = false` and `debug_level = error`.

Reply:
Hm...I guess that is true. Seems overly complicated, but I'm sure you have a plan :)

Reply:
`output_iteration_info defaults` to `false` and maybe we should make `debug_level` default to `error` rather than `none`. Yeah, I think I'll go with that. Then the user will get a reasonable short summary in case of errors and will have to actively enable the tousand-line-dump for the history or actively shush MOOSE completely.

Reply:
👍

Reply:
That plan makes sense to me.